### ISSUE TITLE
Change default script language to expressions

### ISSUE THREAD

**Issue by rjernst** [elastic/elasticsearch#10491]
Since we have non sandboxable dynamic scripts disabled by default, and the sandbox is being removed for groovy (#10480), I think we should change the default language to expression scripts.  This will at least be a better behavior out-of-the-box for users with simple scripts (e.g. #10474).

We should also consider moving groovy back to a separate plugin (separate issue, but just mentioning here). In order to make expressions the default, we will need to make the lucene expression deps non-optional (and possibly shade in asm so as not to guard against conflicts).  Having groovy and expressions in the default setup would then seem overkill (at least at some point in the future, when expressions become more powerful).

**Comment by dakrone** [comment]
+1 for moving to expressions by default

> We should also consider moving groovy back to a separate plugin

I'm -1 on this, because it's still nice to have a "full-power" scripting language that doesn't require doing a node restart to use (just stick some scripts on disk and they are usable)

**Comment by dakrone** [comment]
To clarify, I'm only -1 on removing Groovy if Expressions doesn't support everything we need, if we get it to the point that it does, I think removing Groovy to make it a plugin would be totally fine.

**Comment by javanna** [comment]
I agree with @dakrone .  The only reason why I am not totally sold on making expressions the default is that it can't do updates, but I see how it would at least work out-of-the-box for other usecases without requiring to use file scripts.

**Comment by rjernst** [comment]
I think making it work out of the box is more important than extra work (changing the default) for those that already have to do work to enable scripts (either by writing them on each host or enabling dynamic).

**Comment by clintongormley** [comment]
I don't think we should make another change to the default scripting language until we have a proper replacement which works across the board.  Closing this for now.

**Comment by rmuir** [comment]
I reopened the issue. Groovy is not even working with java 9. I think we should reconsider.

**Comment by clintongormley** [comment]
Java 9 is due to be released in September 2016. I hope to have https://github.com/elastic/elasticsearch/issues/13084 in place long before that. Once merged, we can give it some time to iron out the issues, then switch to that as the default.  

I don't see the point in switching the default to expressions, and then switching again a couple of months later.

**Comment by clintongormley** [comment]
@rmuir I think you said to me that you wanted to move groovy out to its own module, in order to clean up some security issues in core? That work can continue, without us changing the default scripting language for now, no?

**Comment by rmuir** [comment]
I want to move all scripting engines (including expressions) out of the core into plugins. This is necessary due to the security issues around classloaders. 

Personally I think, if we want to bundle certain plugins with our distribution by default, that's what we should do (e.g. distro build will bin/plugin install ones before zipping or whatever), but we keep the separation this way.

**Comment by dadoonet** [comment]
Once we will split core in modules it will be super easy to decide if a module is embedded or exposed as a plugin.

**Comment by clintongormley** [comment]
@rmuir ++

**Comment by rmuir** [comment]
I am not going to wait on some big modularization of the core. I am going to proceed with fixing the security issues around scripting in master.

**Comment by clintongormley** [comment]
Now that we have a new safe scripting language merged into master, which will become the new default as soon as it is ready, we can close this issue.
